### ISSUE TITLE
Context lost and request hangs when uncaughtException happens

### ISSUE THREAD

**Issue by rwky** [getsentry/raven-node#266]
I noticed when using express some errors were missing context (i.e. the url and IP).

After some investigation I've found it's due to uncaughtExceptions, this happens on any POST request that uses the body-parser middleware.

I've narrowed it down to a small test case 

```js
var app = require('express')()
var Raven = require('raven')

Raven.config('config_url').install()
app.use(Raven.requestHandler())

app.use(function(req, res, next) { 
    req.on('data', function() {})
    req.on('end', function(){ next() })
})

app.post('/broken', function (req, res) {
    process.nextTick(function(){
        throw new Error('Broken!')
    })
})

app.post('/works', function(req, res) {
    throw new Error('Works!')
});

app.use(Raven.errorHandler())

app.use(function (err, req, res, next) {
    res.statusCode = 500
    console.log('Sentry ' + res.sentry)
    res.end('Sentry ' + res.sentry + '\n')
})

app.listen(4000)

```
If you run `curl http://localhost:4000/works -d 'test'` it'll behave as expected, if you run `curl http://localhost:4000/broken -d 'test'` the reported error will miss the context **and** the request won't complete since express error handler is never called.

Due to the combination of the `req.on('end', ...` listener and the `process.nextTick` `/broken` generates and `uncaughtException`

Ideally I'd want sentry to log the exception with context, call the error handler so the request ends and then allow me to cleanup (using the callback to install()) which could then terminate the process gracefully.

Does this sound possible?

**Comment by LewisJEllis** [comment]
> Due to the combination of the req.on('end', ... listener and the process.nextTick /broken generates and uncaughtException

Aha! What's happening here is kinda crafty, but you've given nearly the perfect repro case to point me at the issue. Thanks for the great report @rwky.

This is all just an initial look and I haven't verified the repro/potential fix yet, but your repro and description of observed behavior are very consistent are very consistent with what I think is going on. This indeed will only happen when `next` (or some other continuation) is invoked by an event listener, and further only when that event listener is on an event emitter that existed *before* the request-scoped domain (which our express middleware uses to associate context) is created. In this case, we have the continuation `next` invoked by a listener attached to `req` which existed before our `requestHandler` middleware creates the context domain.

It boils down to a trickiness of implicit/explicit binding and how domains work, but ultimately I think this should be fixable with a small change to our `requestHandler` middleware to explicitly bind `req` (and maybe `res` too, unsure if needed) to the context domain. I just need to verify the fix and make sure it won't have other impacts.

**Comment by rwky** [comment]
Glad I could help, it took a while to figure out exactly what was going on, I wanted to make as simple of repro case as possible (also to make sure it wasn't something weird in my own code or middleware).

Just a FYI before we switched to using raven/sentry we used domains to catch errors, log them, kill of the request and process. Each domain had req, res and app bound to it. Similar to this https://nodejs.org/api/domain.html#domain_explicit_binding

I don't think you particularly need app bound but res and res would be good.

**Comment by LewisJEllis** [comment]
Yea, I'm not going to worry about binding app. I'm not even certain that we can guarantee having a reference to the app object - maybe there's one on the req/res object somewhere, but I'm not sure and don't think it's worth investigating. There's only one event ever emitted from the app object and only in a special use case not affecting the average user (and even in that use case I don't think it matters).

Added a PR for this in #269, take a look. Once it and 267 merge I'll release a new version.

**Comment by rwky** [comment]
Agreed on app, the user can always reference app to req/res themselves. Thanks for the fix!

**Comment by LewisJEllis** [comment]
Closed by #269
